src/reducetensor.cpp

```diff
@@ -0,0 +1,619 @@
+    case miopenBFloat16: return (static_cast<int>('B'));
+    case miopenInt8:
+    case miopenInt8x4:
+    case miopenInt32: return (static_cast<int>('O'));
+    default: MIOPEN_THROW("Only float, half, bfloat16 data type is supported."); break;
+    };
+};
+
+inline int GetReduceTensorOpId(miopenReduceTensorOp_t t)
+{
+    switch(t)
+    {
+    case MIOPEN_REDUCE_TENSOR_ADD:
+        return (656868); // 'A' * 10000 + 'D' * 100 + 'D'
+    case MIOPEN_REDUCE_TENSOR_MUL:
+        return (778576); // 'M' * 10000 + 'U' * 100 + 'L'
+    case MIOPEN_REDUCE_TENSOR_MIN:
+        return (777378); // 'M' * 10000 + 'I' * 100 + 'N'
+    case MIOPEN_REDUCE_TENSOR_MAX:
+        return (776588); // 'M' * 10000 + 'A' * 100 + 'X'
+    default: MIOPEN_THROW("Only Add, Mul, Min, Max is supported."); break;
+    };
+};
+
+}; // end of namespace detail
+
+ReduceTensorDescriptor::ReduceTensorDescriptor(miopenReduceTensorOp_t reduceTensorOp,
+                                               miopenDataType_t reduceTensorCompType,
+                                               miopenNanPropagation_t reduceTensorNanOpt,
+                                               miopenReduceTensorIndices_t reduceTensorIndices,
+                                               miopenIndicesType_t reduceTensorIndicesType)
+    : reduceTensorOp_(reduceTensorOp),
+      reduceTensorCompType_(reduceTensorCompType),
+      reduceTensorNanOpt_(reduceTensorNanOpt),
+      reduceTensorIndices_(reduceTensorIndices),
+      reduceTensorIndicesType_(reduceTensorIndicesType)
+{
+    if(reduceTensorIndices == MIOPEN_REDUCE_TENSOR_FLATTENED_INDICES &&
+       reduceTensorIndicesType != MIOPEN_32BIT_INDICES)
+        MIOPEN_THROW("Only int32 type is supported for ReduceTensor indices.");
+};
+
+// return the size of the workspace in bytes, so that the workspace buffer can be prepared by the
+// user
+std::size_t ReduceTensorDescriptor::GetWorkspaceSize(const Handle& handle,
+                                                     const TensorDescriptor& inDesc,
+                                                     const TensorDescriptor& outDesc) const
+{
+    const auto& inDescLengths  = inDesc.GetLengths();
+    const auto& outDescLengths = outDesc.GetLengths();
+
+    if(inDescLengths.size() != outDescLengths.size())
+        MIOPEN_THROW("The number of dimensions of the input and output tensor should match.");
+
+    for(int i = 0; i < inDescLengths.size(); i++)
+    {
+        if(outDescLengths[i] != 1 && outDescLengths[i] != inDescLengths[i])
+            MIOPEN_THROW("The length of the output tensor dimension should either be 1 or be equal "
+                         "to the length of the corresponding dimension of the input tensor.");
+    };
+
+    auto invariantLength = outDesc.GetElementSize();
+    auto toReduceLength  = inDesc.GetElementSize() / invariantLength;
+
+    detail::ReductionKernelConfigurator configurator(256, handle.GetWavefrontWidth());
```

Comment:
@qianfengz @JehandadKhan Currently GetWavefrontWidth returns fixed values of 64, which is incorrect for gfx10+ where warp size is 32. I am going to fix this. How this may affect the `ReductionKernelConfigurator`?

Reply:
The passed the value of `wavefrontsize` to `ReductionKernelConfigurator` will affect the selection of what kernel to use and the sizes of the kernel grid, which would finally bring some variation to the performance, but the effect is not important.   Reduction in MIOpen (so far) does not have performance considered since it does not support tunable for enabling coalesced and vectored data load for data accessing from global memory